engine/engine_test.go

```diff
@@ -2948,6 +2959,97 @@ func TestEngineRecoversFromPanic(t *testing.T) {
 
 }
 
+func TestNativeHistogram(t *testing.T) {
+	opts := promql.EngineOpts{
+		Timeout:              1 * time.Hour,
+		MaxSamples:           1e10,
+		EnableNegativeOffset: true,
+		EnableAtModifier:     true,
+	}
+
+	cases := []struct {
+		name  string
+		query string
+	}{
+		{
+			name:  "plain selector",
+			query: "native_histogram_series",
+		},
+		{
+			name:  "irate() with native histogram",
+			query: "rate(native_histogram_series[1m])",
+		},
+		{
+			name:  "rate() with native histogram",
+			query: "rate(native_histogram_series[1m])",
+		},
+		{
+			name:  "increase() with native histogram",
+			query: "increase(native_histogram_series[1m])",
+		},
+		{
+			name:  "delta() with native and counter histogram",
+			query: "delta(native_histogram_series[1m])",
+		},
+	}
+
+	mixedTypesOpts := []bool{false, true}
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			for _, withMixedTypes := range mixedTypesOpts {
+				t.Run(fmt.Sprintf("mixedTypes=%t", withMixedTypes), func(t *testing.T) {
+					test, err := promql.NewTest(t, "")
+					testutil.Ok(t, err)
+					defer test.Close()
+
+					// Building native histograms
+					lbls := []string{labels.MetricName, "native_histogram_series", "foo", "bar"}
```

Comment:
Question, can we extract the logic of generating native histogram series to a separate function?